### ISSUE TITLE
How I can use PyOD with Image and Text data along with other information present in Tabular data?

### ISSUE THREAD

**Issue by karndeepsingh** [yzhao062/pyod#443]
Hi,
I have data of products and for each product, I have images, descriptions (text data), and product attributes like color, size, etc. How can I use PyOD to do anomaly detection with such mix data? And also what would be right step to convert the data in right format so that I can use PyOD with such mix data to perform Anomaly detection.

Thanks

**Comment by andrewm4894** [comment]
I'd imagine youd need to focus on some sort of multimodel feature representation in your pipeline before you then train a model using PyOD. 

https://en.wikipedia.org/wiki/Multimodal_learning

[This course](https://www.youtube.com/watch?v=VIq5r7mCAyw) is a good place to learn more. 

I dont really think there is any standard way to deal with mixed data like this yet as far as i know (i may be wrong). Is an interesting area of research though. 

It might be more practical to just focus on one type of input data first and start with that. Another idea could be seperate anomaly detection models for each type of data and then just sort of ensemble them in some way. It really can depend on your use case though.

**Comment by karndeepsingh** [comment]
> I'd imagine youd need to focus on some sort of multimodel feature representation in your pipeline before you then train a model using PyOD.
> 
> https://en.wikipedia.org/wiki/Multimodal_learning
> 
> [This course](https://www.youtube.com/watch?v=VIq5r7mCAyw) is a good place to learn more.
> 
> I dont really think there is any standard way to deal with mixed data like this yet as far as i know (i may be wrong). Is an interesting area of research though.
> 
> It might be more practical to just focus on one type of input data first and start with that. Another idea could be seperate anomaly detection models for each type of data and then just sort of ensemble them in some way. It really can depend on your use case though.

Thanks for replying @andrewm4894 

I have trained multimodal using text and image data also thinking to add other categorical attributes so that I can use embeddings generated by this multimodal for anomaly detection.

But once i have these embedddings, how I should prepare this embeddings to use PyOD framework? 
What is the formate PyOD accept the data to be prepared ? Is it always need to be tabular data formate?
also, since I don't have labels for anomalies hence I can't use any supervised method to train the anomaly model. Any suggestion to which type of model I should go for? I was thinking to go for unsupervised models? What would you recommend?

Thanks

**Comment by andrewm4894** [comment]
Lets say each observation is an image and some text. 

You could train some embedding models for each - so lets assume now for the image you have 10 numbers to represent the image in embedding sapce and for the text you also have 10 numbers in the text embedding representation (im just making this up as one example approach). 

For each observation you now have 20 numbers that is your combined representation in some sense.

In PyOD you can train an unsupervised anomaly detection model on some training data and it will try and learn what those 20 numbers typically look like in your training data (no labels required). 

Then at prediiction time if the tyrained model sees something very different to what it has learned as normal that observation would have a higher anomaly score. 

Here is a tutorial i made a while back that goes into detail: https://andrewm4894.com/2020/12/02/anomaly-detection-tutorial/

Although some of the PyOD examples might be even easier to just play around with: https://pyod.readthedocs.io/en/latest/example.html

**Comment by karndeepsingh** [comment]
> 

Thanks again @andrewm4894  for this amazing answer. I was thinking on the same line as you described it. 
I just have a few more questions:
1. Apart from image and text data I also have some more categorical columns. How I can convert these categorical columns to embeddings? I can do some categorical encoding but I think that won't be good. What do you suggest?

2. How I can utilize the semi-supervised approach?  I read the "ADbench" paper which clearly mentioned few label data can help the sem-supervised model to get more accurate on anomaly detection compared to unsupervised.

**Comment by karndeepsingh** [comment]
> Lets say each observation is an image and some text.
> 
> You could train some embedding models for each - so lets assume now for the image you have 10 numbers to represent the image in embedding sapce and for the text you also have 10 numbers in the text embedding representation (im just making this up as one example approach).
> 
> For each observation you now have 20 numbers that is your combined representation in some sense.
> 
> In PyOD you can train an unsupervised anomaly detection model on some training data and it will try and learn what those 20 numbers typically look like in your training data (no labels required).
> 
> Then at prediiction time if the tyrained model sees something very different to what it has learned as normal that observation would have a higher anomaly score.
> 
> Here is a tutorial i made a while back that goes into detail: https://andrewm4894.com/2020/12/02/anomaly-detection-tutorial/
> 
> Although some of the PyOD examples might be even easier to just play around with: https://pyod.readthedocs.io/en/latest/example.html


Thanks again @andrewm4894  for this amazing answer. I was thinking on the same line as you described it. 
I just have a few more questions:
1. Apart from image and text data I also have some more categorical columns. How I can convert these categorical columns to embeddings? I can do some categorical encoding but I think that won't be good. What do you suggest?

2. How I can utilize the semi-supervised approach?  I read the "ADbench" paper which clearly mentioned few label data can help the sem-supervised model to get more accurate on anomaly detection compared to unsupervised.

**Comment by karndeepsingh** [comment]
> > Lets say each observation is an image and some text.
> > You could train some embedding models for each - so lets assume now for the image you have 10 numbers to represent the image in embedding sapce and for the text you also have 10 numbers in the text embedding representation (im just making this up as one example approach).
> > For each observation you now have 20 numbers that is your combined representation in some sense.
> > In PyOD you can train an unsupervised anomaly detection model on some training data and it will try and learn what those 20 numbers typically look like in your training data (no labels required).
> > Then at prediiction time if the tyrained model sees something very different to what it has learned as normal that observation would have a higher anomaly score.
> > Here is a tutorial i made a while back that goes into detail: https://andrewm4894.com/2020/12/02/anomaly-detection-tutorial/
> > Although some of the PyOD examples might be even easier to just play around with: https://pyod.readthedocs.io/en/latest/example.html
> 
> Thanks again @andrewm4894 for this amazing answer. I was thinking on the same line as you described it. I just have a few more questions:
> 
> 1. Apart from image and text data I also have some more categorical columns. How I can convert these categorical columns to embeddings? I can do some categorical encoding but I think that won't be good. What do you suggest?
> 2. How I can utilize the semi-supervised approach?  I read the "ADbench" paper which clearly mentioned few label data can help the sem-supervised model to get more accurate on anomaly detection compared to unsupervised.

**Comment by karndeepsingh** [comment]
> https://github.com/yzhao062/pyod/tree/master/examples

Thank you @yzhao062 for creating such an awesome framework. I have read a paper on "ADBench" and I have a few questions about it:
 In the paper, it was mentioned that the selection of the algorithm depends upon the type of anomaly. How we can understand the type of anomaly present in our data when we have "image", "text" and "other categorical columns"? can you suggest some ways to detect the type of anomaly for such mix type of data? 

Thanks
